backend/lucene/src/main/java/org/hibernate/search/backend/lucene/search/predicate/impl/LuceneKnnPredicate.java

```diff
@@ -31,23 +33,29 @@ public class LuceneKnnPredicate extends AbstractLuceneSingleFieldPredicate imple
 	private final int k;
 	private final Object vector;
 	private final LuceneSearchPredicate filter;
+	private final Float similarity;
+	private final VectorSimilarityFunction similarityFunction;
 
 	private LuceneKnnPredicate(Builder<?> builder) {
 		super( builder );
 		this.k = builder.k;
 		this.vector = builder.vector;
 		this.filter = builder.filter;
+		this.similarity = builder.similarity;
+		this.similarityFunction = builder.similarityFunction;
 	}
 
 	@Override
 	protected Query doToQuery(PredicateRequestContext context) {
 		if ( vector instanceof byte[] ) {
-			return new KnnByteVectorQuery(
+			KnnByteVectorQuery query = new KnnByteVectorQuery(
 					absoluteFieldPath, (byte[]) vector, k, filter == null ? null : filter.toQuery( context ) );
+			return similarity == null ? query : VectorSimilarityFilterQuery.create( query, similarity, similarityFunction );
```

Comment:
Isn't there a way to plug into the `Knn*Query`, so that you filter out the documents *while collecting them*? That could improve performance significantly in some edge cases (large `k`, high `minSimilarity`,  ...).

Reply:
I didn't see any, and from what I've seen the collecting part lives somewhere at the reader level when searching for the vectors, but I'll double-check 🙈